### PR TITLE
Collada reader writer improvement for webstack2

### PR DESCRIPTION
**Pull request by ziyan**
Included in this PR:

1. Some more minor `_mutexInterfaces` locking fix in `environment-core.h`
2. Add `mustresolveuri` option to collada reader and JSON reader, to throw exception when referenced scene is not found
3. Keep articulate system id in collada stable, so that when we convert `body2_motion` from json to collada, it remains `body2_motion`, otherwise it becomes `body0_motion` and other **dae** scene referring to `body2_motion` will not resolve successfully
4. Keep link id and joint id in collada stable, so that when we convert a scene that has external reference, those external references can be correctly resolved
5. remove auto assignment of id in openrave, leave them empty, save json as-is

  | Scenario | Body ID | Link ID | Joint ID | Geometry ID
-- | -- | -- | -- | -- | --
1 | robotbridge or planning slave load scene from COLLADA | take articulated system id | take link sid | take joint sid | generate as geomN
2 | robotbridge or planning slave load scene from JSON | as-is in file, could be empty | as-is in file, could be empty | as-is in file, could be empty | as-is in file, could be empty
3 | slave load dynamic objects from COLLADA | generate random body id | take link sid | take joint sid | generate as geomN2
4 | slave load dynamic objects from JSON | generate random body id | as-is in file, could be empty | as-is in file, could be empty | as-is in file, could be empty
5 | save env to COLLADA | try to reuse bodyX_motion, generate unique numeric id | use info id or generate random one | use info id or generate random one | generate sequential id
6 | save env to JSON | as-is, save to file, with empty id | as-is, save to file, with empty id | as-is, save to file, with empty id | as-is, save to file, with empty id
7 | construct body at runtime | empty | empty | empty | empty

### CODE REVIEW

File: src/libopenrave-core/colladaparser/colladawriter.cpp [hunk 1]
```diff
@@ -516,8 +516,29 @@ class ColladaWriter : public daeErrorHandler
...
+            const std::string& bodyId = (*itbody)->_id;
+            // try to restore what id was used
+            // check for id that has format bodyX_motion
+            if (bodyId.size() > 11 && bodyId.substr(0, 4) == "body" && bodyId.substr(bodyId.size()-7) == "_motion") {
+                try {
+                    int numericBodyId = boost::lexical_cast<int>(bodyId.substr(4, bodyId.size()-11));
+                    if (numericBodyId >= globalid) {
+                        globalid = numericBodyId+1;
+                    }
+                    _mapBodyIds[(*itbody)->GetEnvironmentId()] = numericBodyId;
```
Comment (rdiankov) [inline src/libopenrave-core/colladaparser/colladawriter.cpp:528]: @ziyan how do you know if `numericBodyId` is unique? shouldn't you check in `_mapBodyIds`?

File: src/libopenrave-core/colladaparser/colladareader.cpp [hunk 2]
```diff
@@ -3672,10 +3680,16 @@ class ColladaReader : public daeErrorHandler
                         pbody = robots.front();
                     } else {
                         RAVELOG_DEBUG_FORMAT("Found %d robots, Do not support this case for uri %s", robots.size() % uri);
+                        if (_bMustResolveURI) {
+                            throw OPENRAVE_EXCEPTION_FORMAT("failed to load connected body from uri \"%s\", more than one robot found", uri, ORE_InvalidArguments);
```
Comment (rdiankov) [inline src/libopenrave-core/colladaparser/colladareader.cpp:3684]: @ziyan let's have another `ORE_` code for URI not found. thanks

File: src/libopenrave-core/colladaparser/colladawriter.cpp (hunk 1 above)
Comment (ziyan) [reply src/libopenrave-core/colladaparser/colladawriter.cpp:528]: hmm, I am assuming `_id` is unique, and `_mapBodyIds` at this point is empty, but I can check.

File: src/libopenrave-core/colladaparser/colladareader.cpp [hunk 3]
```diff
@@ -898,6 +903,11 @@ class ColladaReader : public daeErrorHandler
             }
         }
 
+        if( bSuccess ) {
+            // for extracting a single body from file, need to assign a random body id to not conflict with what's already in environment
+            probot->_id = str(boost::format("body%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/colladaparser/colladareader.cpp:908]: @ziyan random values like this are not a good idea. Please change the implementation so the same consistent value is produced loading the same scene, no matter when or how many times i load that scene.

File: src/libopenrave-core/colladaparser/colladareader.cpp [hunk 4]
```diff
@@ -967,6 +977,12 @@ class ColladaReader : public daeErrorHandler
                 break;
             }
         }
+
+        if( bSuccess ) {
+            // for extracting a single body from file, need to assign a random body id to not conflict with what's already in environment
+            pbody->_id = str(boost::format("body%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/colladaparser/colladareader.cpp:983]: @ziyan same as above

File: src/libopenrave-core/colladaparser/colladareader.cpp [hunk 5]
```diff
@@ -988,6 +1004,12 @@ class ColladaReader : public daeErrorHandler
         if( bSuccess && _prefix.size() > 0 ) {
             _AddPrefixForKinBody(pbody,_prefix);
         }
+
+        if( bSuccess ) {
+            // for extracting a single body from file, need to assign a random body id to not conflict with what's already in environment
+            pbody->_id = str(boost::format("body%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/colladaparser/colladareader.cpp:1010]: @ziyan same as above

File: src/libopenrave-core/colladaparser/colladawriter.cpp [hunk 6]
```diff
@@ -2779,14 +2804,37 @@ class ColladaWriter : public daeErrorHandler
...
+            // check duplicate
+            FOREACHC(itId, _mapBodyLinkIds[plink->GetParent()->GetEnvironmentId()]) {
+                if (itId->second == linksid) {
+                    linksid.clear();
+                    break;
+                }
+            }
+        }
+        if (linksid.empty()) {
+            linksid = str(boost::format("link%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/colladaparser/colladawriter.cpp:2834]: @ziyan same as above

File: src/libopenrave-core/colladaparser/colladawriter.cpp [hunk 7]
```diff
@@ -2800,18 +2848,30 @@ class ColladaWriter : public daeErrorHandler
...
+                if (itId->second == jointsid) {
+                    jointsid.clear();
                     break;
                 }
-                ++index;
             }
         }
-        return str(boost::format("node_joint%d_axis%d")%index%iaxis);
+        if (jointsid.empty()) {
+            jointsid = str(boost::format("joint%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/colladaparser/colladawriter.cpp:2867]: @ziyan same as above

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 8]
```diff
@@ -469,6 +477,8 @@ class JSONReader
             if( !pRobot ) {
                 return false;
             }
+            // for extracting a single body from file, need to assign a random body id to not conflict with what's already in environment
+            pRobotBaseInfo->_id = str(boost::format("body%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:481]: @ziyan same as above

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 9]
```diff
@@ -481,6 +491,8 @@ class JSONReader
             if( !pBody ) {
                 return false;
             }
+            // for extracting a single body from file, need to assign a random body id to not conflict with what's already in environment
+            pKinBodyInfo->_id = str(boost::format("body%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:495]: @ziyan same as above

File: src/libopenrave-core/jsonparser/jsonreader.cpp [hunk 10]
```diff
@@ -510,6 +521,8 @@ class JSONReader
         if( !pRobot ) {
             return false;
         }
+        // for extracting a single body from file, need to assign a random body id to not conflict with what's already in environment
+        pRobotBaseInfo->_id = str(boost::format("body%d")%utils::GetNanoTime());
```
Comment (rdiankov) [inline src/libopenrave-core/jsonparser/jsonreader.cpp:525]: @ziyan same as above

File: src/libopenrave-core/colladaparser/colladareader.cpp (hunk 3 above)
Comment (ziyan) [reply src/libopenrave-core/colladaparser/colladareader.cpp]: For now, I think empty id is okay for this case. Resolved.

File: src/libopenrave-core/colladaparser/colladareader.cpp (hunk 4 above)
Comment (ziyan) [reply src/libopenrave-core/colladaparser/colladareader.cpp]: For now, I think empty id is okay for this case. Resolved.

File: src/libopenrave-core/colladaparser/colladareader.cpp (hunk 5 above)
Comment (ziyan) [reply src/libopenrave-core/colladaparser/colladareader.cpp]: For now, I think empty id is okay for this case. Resolved.

File: src/libopenrave-core/colladaparser/colladawriter.cpp (hunk 6 above)
Comment (ziyan) [reply src/libopenrave-core/colladaparser/colladawriter.cpp]: Provided stable sid.

File: src/libopenrave-core/colladaparser/colladawriter.cpp (hunk 7 above)
Comment (ziyan) [reply src/libopenrave-core/colladaparser/colladawriter.cpp]: Provided stable sid.

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 8 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsonreader.cpp]: For now, I think empty id is okay for this case. Resolved.

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 9 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsonreader.cpp]: For now, I think empty id is okay for this case. Resolved.

File: src/libopenrave-core/jsonparser/jsonreader.cpp (hunk 10 above)
Comment (ziyan) [reply src/libopenrave-core/jsonparser/jsonreader.cpp]: For now, I think empty id is okay for this case. Resolved.

File: src/libopenrave/robot.cpp [hunk 11]
```diff
@@ -2747,6 +2537,10 @@ UpdateFromInfoResult RobotBase::UpdateFromRobotInfo(const RobotBaseInfo& info)
 
     // manipulators
     FOREACHC(itManipulatorInfo, info._vManipulatorInfos) {
+        if ((*itManipulatorInfo)->_id.empty()) {
```
Comment (rdiankov) [inline src/libopenrave/robot.cpp:2540]: @ziyan skip updating because no `id`? is that safe? perhaps we should match names..?

Comment (ziyan) [reply src/libopenrave/robot.cpp]: No longer skipping id.

**Comment by rdiankov** [comment]
woot~

**Pull request merged**